<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Leighton Pritchard">

<title>1&nbsp; Introduction – BM329 Workshop D: Diseases of the Respiratory Tract - Supporting Information</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./upgma.html" rel="next">
<link href="./index.html" rel="prev">
<link href="./icon_32.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-a34d670291f06f286357e447776a572a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./intro.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">BM329 Workshop D: Diseases of the Respiratory Tract - Supporting Information</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/sipbs-compbiol/BM329_Block_A_workshop" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface to the 2024-25 BM329 Block D workshop</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./upgma.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">UPGMA</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Summary</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./additional_files.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional Files</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#dendrograms-tree-diagrams-and-evolution" id="toc-dendrograms-tree-diagrams-and-evolution" class="nav-link active" data-scroll-target="#dendrograms-tree-diagrams-and-evolution"><span class="header-section-number">1.1</span> Dendrograms (Tree Diagrams) and Evolution</a></li>
  <li><a href="#interpreting-a-tree" id="toc-interpreting-a-tree" class="nav-link" data-scroll-target="#interpreting-a-tree"><span class="header-section-number">1.2</span> Interpreting a tree</a>
  <ul class="collapse">
  <li><a href="#branch-lengths" id="toc-branch-lengths" class="nav-link" data-scroll-target="#branch-lengths"><span class="header-section-number">1.2.1</span> Branch lengths</a></li>
  <li><a href="#cladograms-and-phylograms" id="toc-cladograms-and-phylograms" class="nav-link" data-scroll-target="#cladograms-and-phylograms"><span class="header-section-number">1.2.2</span> Cladograms and Phylograms</a></li>
  <li><a href="#topology" id="toc-topology" class="nav-link" data-scroll-target="#topology"><span class="header-section-number">1.2.3</span> Topology</a></li>
  </ul></li>
  <li><a href="#evolution-and-clustering" id="toc-evolution-and-clustering" class="nav-link" data-scroll-target="#evolution-and-clustering"><span class="header-section-number">1.3</span> Evolution and Clustering</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/sipbs-compbiol/BM329_Block_A_workshop/edit/main/intro.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/sipbs-compbiol/BM329_Block_A_workshop/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This workshop introduces and reinforces concepts related to transmission and evolution of respiratory viruses.</p>
<section id="dendrograms-tree-diagrams-and-evolution" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="dendrograms-tree-diagrams-and-evolution"><span class="header-section-number">1.1</span> Dendrograms (Tree Diagrams) and Evolution</h2>
<p>In this workshop, you will produce <a href="https://en.wikipedia.org/wiki/Dendrogram"><em>dendrograms</em></a>: these are <em>tree diagrams</em> that can represent the process of evolution. You have already seen tree diagrams, like those in <a href="#fig-tol-rect" class="quarto-xref">Figure&nbsp;<span>1.1</span></a>, used on this course to represent the process of evolution.</p>
<div id="fig-tol-rect" class="quarto-float quarto-figure quarto-figure-center anchored" alt="Rectangular phylogram of the Tree of Life. Archaea (in red) and Bacteria (in blue) form the two main domains, and Eukaryotes (in green) are a subgroup within Archaea.">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-tol-rect-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/41564_2016_Article_BFnmicrobiol201648_Fig2_HTML.png" class="img-fluid figure-img" style="width:80.0%" alt="Rectangular phylogram of the Tree of Life. Archaea (in red) and Bacteria (in blue) form the two main domains, and Eukaryotes (in green) are a subgroup within Archaea.">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-tol-rect-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1.1: Tree of Life obtained using ribosomal genes, showing the major Domains of life. Reproduced from <span class="citation" data-cites="Hug2016-sv">Hug et al. (<a href="references.html#ref-Hug2016-sv" role="doc-biblioref">2016</a>)</span>.
</figcaption>
</figure>
</div>
<p>Typically, in evolutionary analyses, you will see that these trees are organised to show a progression through time. The <em>leaves</em> of the tree usually represent things that exist “now” or, at least, most recently (these are at the right hand side of <a href="#fig-tol-rect" class="quarto-xref">Figure&nbsp;<span>1.1</span></a>).</p>
<p>The dendrogram traces lines - <em>branches</em> - from the leaves, and these gradually meet up together, just as branches of a real tree do, as you progress from right to left in <a href="#fig-tol-rect" class="quarto-xref">Figure&nbsp;<span>1.1</span></a>. Eventually, they all meet up together at the oldest (left-most, in <a href="#fig-tol-rect" class="quarto-xref">Figure&nbsp;<span>1.1</span></a>) part of the tree, (called the <em>root</em>).</p>
<p>Trees like this can be used to represent large-scale evolution, like the complete Tree of Life in <a href="#fig-tol-rect" class="quarto-xref">Figure&nbsp;<span>1.1</span></a>, or evolution on a small scale. For example, <a href="#fig-pseud-hosp" class="quarto-xref">Figure&nbsp;<span>1.2</span></a> represents evolution of the pathogenic bacterium <em>Pseudomonas aeruginosa</em> in a UK hospital. This tree was used to trace the source of infection in a burns unit, ultimately locating the precise valve in the plumbing from which the infectious agent was finding its way into patients (<span class="citation" data-cites="Quick2014-xp">Quick et al. (<a href="references.html#ref-Quick2014-xp" role="doc-biblioref">2014</a>)</span>).</p>
<div id="fig-pseud-hosp" class="quarto-float quarto-figure quarto-figure-center anchored" alt="Phylogenetic tree of Pseudomonas aeruginosa isolates, showing a close relationship between those obtained from wound swabs, and those from the shower next to the bed (@Quick2014-xp).">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-pseud-hosp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/bmjopen-2014-November-4-11--F3.large.jpg" class="img-fluid figure-img" style="width:80.0%" alt="Phylogenetic tree of Pseudomonas aeruginosa isolates, showing a close relationship between those obtained from wound swabs, and those from the shower next to the bed (@Quick2014-xp).">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pseud-hosp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1.2: High-resolution genome-based phylogenetic tree of <em>Pseudomonas aeruginosa</em> samples collected during an observational hospital study to track infection. Samples were collected from wounds (red), water (blue), or the environment (green). All wound samples are closely-related to water samples from the shower next to bed 11. Reproduced from <span class="citation" data-cites="Quick2014-xp">Quick et al. (<a href="references.html#ref-Quick2014-xp" role="doc-biblioref">2014</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="interpreting-a-tree" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="interpreting-a-tree"><span class="header-section-number">1.2</span> Interpreting a tree</h2>
<p>The trees in <a href="#fig-tol-rect" class="quarto-xref">Figure&nbsp;<span>1.1</span></a> and <a href="#fig-pseud-hosp" class="quarto-xref">Figure&nbsp;<span>1.2</span></a> both imply a branching evolutionary process. A single common ancestor to every organism represented at one of the leaves existed at some point in history, and that date in history is on the very leftmost point of the tree. The rest of the tree is a representation of how evolution progressed from that single ancestor to the variety of organisms represented at the <em>leaf nodes</em>.</p>
<section id="branch-lengths" class="level3" data-number="1.2.1">
<h3 data-number="1.2.1" class="anchored" data-anchor-id="branch-lengths"><span class="header-section-number">1.2.1</span> Branch lengths</h3>
<p>As you move from left-to-right in those trees, there is a short horizontal line (the <em>root</em>) that represents time passing for that ancestor. But then the branch splits into two (it <em>diverges</em>), representing some kind of event that produces distinct “offspring.” Each of these proceeds through time as a horizontal line for a distance before itself <em>diverging</em> to result in two new, differentiable offspring, and so on and so on until the <em>leaf nodes</em> are reached.</p>
<p>The lengths of the branches here represent, in some way, the passage of time between consecutive <em>divergences</em> (or a leaf node and the most recent split). The longer the branch, the longer the passage of time.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The length of a branch doesn’t necessarily correspond exactly to time
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>In practice, when we make these trees, we are usually measuring something other than time, itself. More often we are measuring some kind of difference between the leaf nodes, and <em>assuming</em> that a difference of 200 corresponds to 200 units of time (years, millennia, etc.), and a difference of 1000 corresponds to 1000 units.</p>
<p>Biology is complicated, and this assumption rarely holds exactly. There are methods to try to turn our measurements accurately into units of time, but they are beyond the scope of this course.</p>
</div>
</div>
</div>
</section>
<section id="cladograms-and-phylograms" class="level3" data-number="1.2.2">
<h3 data-number="1.2.2" class="anchored" data-anchor-id="cladograms-and-phylograms"><span class="header-section-number">1.2.2</span> Cladograms and Phylograms</h3>
<p>There is more than one kind of dendrogram. Two types you will meet frequently are:</p>
<ul>
<li><strong><em>phylograms</em></strong>: each branch length is intended to represent the passage of time (or some other measure of difference) - they represent change on an evolutionary timescale</li>
<li><strong><em>cladograms</em></strong>: all branch lengths are the same - they do not represent change on an evolutionary timescale, only the order of divergences in the tree</li>
</ul>
<div id="fig-clad-phyl" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-clad-phyl-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-clad-phyl" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-phylo" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-phylo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/example_tree_phylo.png" class="img-fluid figure-img" data-ref-parent="fig-clad-phyl">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-phylo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Phylogram.
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-clad-phyl" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-clado" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-clado-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/example_tree_clado.png" class="img-fluid figure-img" data-ref-parent="fig-clad-phyl">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-clado-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Cladogram
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-clad-phyl-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1.3: The same evolutionary tree represented as a <em>phylogram</em>, and as a <em>cladogram</em>. In a phylogram the branch lengths may be uneven. They represent evolutionary time between <em>nodes</em> (divergences and/or leaves on the tree), and the time between those may vary. In a cladogram the branch lengths are identical and the tree only represents <em>topology</em> - the order in which branches diverge.
</figcaption>
</figure>
</div>
</section>
<section id="topology" class="level3" data-number="1.2.3">
<h3 data-number="1.2.3" class="anchored" data-anchor-id="topology"><span class="header-section-number">1.2.3</span> Topology</h3>
<p><em>Topology</em> is the order of branching in a tree: the order of <em>divergence events</em> from the common ancestor of all leaf nodes to the present. Two trees with the same order of branching, but different branch lengths, have the same <em>topology</em>, and imply the same sequence of divergences.</p>
<p>For example, in <a href="#fig-clad-phyl" class="quarto-xref">Figure&nbsp;<span>1.3</span></a> the first divergence separates <code>Mouse</code> from all the other mammals, and then the next divergence separates <code>Bovine</code> from the primates. Whether you examine the phylogram or the cladogram, the order of branching - the <em>topology</em> - of the tree is the same. They represent the same evolutionary sequence of events.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Even if two trees share the same topology, and the same <em>sequence</em> of events, if any branch lengths differ the trees might still represent different <em>evolutionary histories</em> as the times between those events may differ.</p>
</div>
</div>
<p>Two trees can look superficially different, but actually represent the same sequence of branching events, and even the same evolutionary history, as is the case in <a href="#fig-phylo-rot" class="quarto-xref">Figure&nbsp;<span>1.4</span></a>. Here, the tree has been rotated around the node that joins <code>Human</code> to <code>Chimp</code>. Although <code>Human</code> is above <code>Chimp</code> in one tree and below it in the other, both trees have identical topology.</p>
<div id="fig-phylo-rot" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-phylo-rot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-phylo-rot" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-phylo-rot1" class="quarto-float quarto-figure quarto-figure-center anchored" alt="Phylogenetic tree of mammals, with the clade containing only Human and Chimp highlighted, Chimp above Human.">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-phylo-rot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/example_tree_phylo_rot_1.png" class="img-fluid figure-img" data-ref-parent="fig-phylo-rot" alt="Phylogenetic tree of mammals, with the clade containing only Human and Chimp highlighted, Chimp above Human.">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-phylo-rot1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Original tree
</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-phylo-rot" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-phylo-rot2" class="quarto-float quarto-figure quarto-figure-center anchored" alt="Phylogenetic tree of mammals, with the clade containing only Human and Chimp highlighted, Human above Chimp.">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-phylo-rot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/example_tree_phylo_rot_2.png" class="img-fluid figure-img" data-ref-parent="fig-phylo-rot" alt="Phylogenetic tree of mammals, with the clade containing only Human and Chimp highlighted, Human above Chimp.">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-phylo-rot2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Tree with rotation
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-phylo-rot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1.4: In any tree, branches can be rotated around an <em>internal node</em> without changing the structure or interpretation of the tree. Here, the two leaf nodes highlighted in purple (<a href="#fig-phylo-rot1" class="quarto-xref">Figure&nbsp;<span>1.4 (a)</span></a>) are rotated around their common ancestor to give the tree in <a href="#fig-phylo-rot2" class="quarto-xref">Figure&nbsp;<span>1.4 (b)</span></a>. The rotated conformation is highlighted in green. Even though the two trees <em>look</em> different, the evolutionary relationships they represent are identical.
</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-important callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Is the tree represented in <a href="#fig-phylo-rot" class="quarto-xref">Figure&nbsp;<span>1.4</span></a> equivalent to the tree shown in <a href="#fig-phylo" class="quarto-xref">Figure&nbsp;<span>1.3 (a)</span></a>?</p>
</div>
</div>
</section>
</section>
<section id="evolution-and-clustering" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="evolution-and-clustering"><span class="header-section-number">1.3</span> Evolution and Clustering</h2>
<p>The examples in <a href="#fig-tol-rect" class="quarto-xref">Figure&nbsp;<span>1.1</span></a>, <a href="#fig-pseud-hosp" class="quarto-xref">Figure&nbsp;<span>1.2</span></a>, and <a href="#fig-phylo-rot" class="quarto-xref">Figure&nbsp;<span>1.4</span></a> are representations of evolution. In all cases, two organisms that share a more recent common ancestor (like <code>Chimp</code> and <code>Human</code> in <a href="#fig-phylo-rot" class="quarto-xref">Figure&nbsp;<span>1.4</span></a>) combine together at a closer <em>node</em> than do two more distantly-related organisms (like <code>Chimp</code> and <code>Gibbon</code>).</p>
<p>It is natural to think of this representation as the divergence of species, as <code>Gibbon</code> diverging from ancestral primates at an earlier stage than <code>Gorilla</code> diverging from the ancestor of <code>Chimp</code> and <code>Human</code>. But it is <em>also a representation of <strong>clustering</strong></em>. What this means is that we <em>cluster</em> <code>Chimp</code> more closely with <code>Human</code> than with <code>Bovine</code> because <code>Chimp</code> and <code>Human</code> share more similarities with each other than either do with <code>Bovine</code>.</p>
<p>In practice this means we can use mathematical approaches that cluster similar things together into trees to approximate evolutionary history, so long as what we use to measure <em>similarity</em> is relevant to evolution. Methods that perform this kind of clustering to produce a tree are called <a href="https://www.learndatasci.com/glossary/hierarchical-clustering/"><strong><em>hierarchical clustering methods</em></strong></a> and you will use one of these to infer an evolutionary history for the `flu viruses you are evolving in the workshop.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Two different ways to reconstruct evolution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><a href="https://nlp.stanford.edu/IR-book/html/htmledition/hierarchical-clustering-1.html">Hierarchical clustering methods</a> are only one way to build an evolutionary tree. They were the first to be used, because they are not difficult to understand, and are relatively straightforward and quick to calculate. All of these methods apply an <em>algorithm</em> to construct a tree (or <em>hierarchy</em> - hence “hierarchical”) from a <em>matrix</em> (table) of distances between organisms. Methods that use this approach include Neighbour-Joining (<span class="citation" data-cites="Gascuel2006-ku">Gascuel and Steel (<a href="references.html#ref-Gascuel2006-ku" role="doc-biblioref">2006</a>)</span>), and <a href="https://www.icp.ucl.ac.be/~opperd/private/upgma.html">UPGMA</a>. For these methods to produce an evolutionary tree, the distance we measure between organisms should reflect their evolutionary separation.</p>
<p>The development of cheap, powerful computing enabled more advanced and more accurate mathematical methods of evolutionary reconstruction to be used routinely. Such methods include Maximum Likelihood (ML, <span class="citation" data-cites="Xia2018-cs">Xia (<a href="references.html#ref-Xia2018-cs" role="doc-biblioref">2018</a>)</span>) and Bayesian (<span class="citation" data-cites="Nascimento2017-nc">Nascimento, Reis, and Yang (<a href="references.html#ref-Nascimento2017-nc" role="doc-biblioref">2017</a>)</span>) approaches. These methods require significant computing power and differ from hierarchical clustering because they attempt to <em>fit</em> statistically a <em>model of evolution</em> (like fitting a curve on a graph) to the data obtained from each organism, rather than using an algorithm to build a tree from distances.</p>
<p>In modern biology, a standard laptop can very quickly produce trees using Maximum Likelihood methods that fit an explicit and well-justified model of evolution. As a result algorithmic methods, such as Neighbour-Joining and UPGMA, which are prone to systematic errors and inaccuracies, and whose only real advantage was speed, are no longer considered good practice for evolutionary reconstruction.</p>
<p>If you’re interested in learning more about how we do this in practice, you might be interested in this course, by Dr Conor Meehan:</p>
<ul>
<li><a href="https://conmeehan.github.io/PathogenDataCourse/IntroToPhylogenetics.html">Introduction to Phylogenetics</a></li>
</ul>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Let's build a tree">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Let’s build a tree
</div>
</div>
<div class="callout-body-container callout-body">
<p>Let’s get started with building a UPGMA tree by clicking on the link <a href="./upgma.html"><code>UPGMA</code></a> (here, in the menu, or below)</p>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-Gascuel2006-ku" class="csl-entry" role="listitem">
Gascuel, Olivier, and Mike Steel. 2006. <span>“Neighbor-Joining Revealed.”</span> <em>Mol. Biol. Evol.</em> 23 (11): 1997–2000.
</div>
<div id="ref-Hug2016-sv" class="csl-entry" role="listitem">
Hug, Laura A, Brett J Baker, Karthik Anantharaman, Christopher T Brown, Alexander J Probst, Cindy J Castelle, Cristina N Butterfield, et al. 2016. <span>“A New View of the Tree of Life.”</span> <em>Nat. Microbiol.</em> 1 (5).
</div>
<div id="ref-Nascimento2017-nc" class="csl-entry" role="listitem">
Nascimento, Fabrı́cia F, Mario dos Reis, and Ziheng Yang. 2017. <span>“A Biologist’s Guide to Bayesian Phylogenetic Analysis.”</span> <em>Nat. Ecol. Evol.</em> 1 (10): 1446–54.
</div>
<div id="ref-Quick2014-xp" class="csl-entry" role="listitem">
Quick, Joshua, Nicola Cumley, Christopher M Wearn, Marc Niebel, Chrystala Constantinidou, Chris M Thomas, Mark J Pallen, et al. 2014. <span>“Seeking the Source of Pseudomonas Aeruginosa Infections in a Recently Opened Hospital: An Observational Study Using Whole-Genome Sequencing.”</span> <em>BMJ Open</em> 4 (11): e006278.
</div>
<div id="ref-Xia2018-cs" class="csl-entry" role="listitem">
Xia, Xuhua. 2018. <span>“Maximum Likelihood in Molecular Phylogenetics.”</span> In <em>Bioinformatics and the Cell</em>, 381–95. Cham: Springer International Publishing.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link" aria-label="Preface to the {{< var ay >}} BM329 Block {{< var block >}} workshop">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface to the 2024-25 BM329 Block D workshop</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./upgma.html" class="pagination-link" aria-label="UPGMA">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">UPGMA</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>These BM329 workshops were written by Dr Leighton Pritchard, Dr Morgan Feeney, and Prof.&nbsp;Paul Hoskisson.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/sipbs-compbiol/BM329_Block_A_workshop/edit/main/intro.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/sipbs-compbiol/BM329_Block_A_workshop/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>